### PR TITLE
Dependency upgrade: xstream 1.4.19, jettison 1.4.1

### PR DESCRIPTION
**Pull request by groldan**
Current XStream version has several (29) [vulnerabilities](https://mvnrepository.com/artifact/com.thoughtworks.xstream/xstream/1.4.11.1), though **they won't affect** GeoServer, since it properly configures the security subsystem, an upgrade seems to be in place anyways.

This upgrades to the latest `1.4.19` version and Jettison `1.4.1` with a patch to preserve Jettison's legacy
behaviour when encoding single-element arrays.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by groldan** [comment]
Supersedes #4572

**Comment by aaime** [comment]
AFAIK the vulnerabilities occur only when the XStream security subsystem is not configured. The GeoServer code does initialize it (although I cannot guarantee), so it should not actually be vulnerable? Communication about potential vulnerabilities should be done in a responsible way, without spreading panic, and done only once fixes are released to the public, please.

**Comment by groldan** [comment]
Build's failing on `CoverageControllerWCSTest.testPostAsJSON()`.

I wonder why `CoverageControllerWCSTest.testPostAsJSON()` expects a single-element-array:

```java
        JSONArray coverages = json.getJSONObject("coverages").getJSONArray("coverage");
        assertThat(coverages.size(), is(1));
```
While `CoverageControllerTest.testGetAsJSON()` expects an object:

```java
        JSONObject coverage = ((JSONObject) json).getJSONObject("coverage");
        assertNotNull(coverage);
```

**Comment by aaime** [comment]
@groldan yes spreading panic, it should be obvious to anyone that has had exposure to customers, which go nuts over the off-chance possibility anything is wrong. The Log4J2 case should serve as an emblematic example, we got grilled over a case of "run a RCE with any HTTP request" when that was not even affecting GeoServer, and ended up fixing one that required pre-existing access to the server file system in order to be triggered. Why that? Because people panic before checking if they are even affected. Even just mentioning "vulnerability" makes some people trigger... and those same people typically don't check CVEs and don't do research (the latter kind normally checks if an issue is an actual one, and how it can be triggered, before taking out the big guns).

We have multiple sections about responsible disclosure:
- http://geoserver.org/comm/
- https://osgeo-org.atlassian.net/projects/GEOS/summary/

Please read them and be more careful in the future, even for issues that are already listed (and even more so for the ones that are not, or for which you have an attack vector), the more they are discussed in public, the more the nutjobs mentioned above will start panicking.

**Comment by aaime** [comment]
> Build's failing on `CoverageControllerWCSTest.testPostAsJSON()`.
> 
> I wonder why `CoverageControllerWCSTest.testPostAsJSON()` expects a single-element-array:
> 
> ```java
>         JSONArray coverages = json.getJSONObject("coverages").getJSONArray("coverage");
>         assertThat(coverages.size(), is(1));
> ```
> 
> While `CoverageControllerTest.testGetAsJSON()` expects an object:
> 
> ```java
>         JSONObject coverage = ((JSONObject) json).getJSONObject("coverage");
>         assertNotNull(coverage);
> ```

I guessing it's a simple "because XStream used to generate outputs like this"... historically XStream was used for the configuration, the REST API came some time after tha, and that's when we also got a JSON encoding... I believe it was accepted as-is, and tests kept on being written using that "untouchable" behavior.
 
In #4572 I could not find a set of settings that made the output match the previous one. As weird as it is, and as much as I don't like the inconsistency, the test spotted a backwards incompatible change. We can have such change, it just requires discussion with the PSC and an indication to the users that the upgrade might break some REST automation. I'll be +1 on breaking the JSON encoding expectations for more correct output, btw.

**Comment by groldan** [comment]
Hi, for the moment I fixed the test to expect an object, under the assumption that since it's already the expected behavior, seems more like a fix than anything else, like in we at least have a consistent output? Let me know what you think.

**Comment by groldan** [comment]
> The write side is more problematic, will the change make the REST API stop accepting JSON it used to read?

Haven't changed anything related to writes (POST requests), relying on the existing set of tests.
The only thing I've changed is what that test expects after the POST was processed, and reads the coverage created by the POST request. For some strange reason it used to expect a single-element array, whereas everywhere else, a GET request with a single element returns an object.

Is that what you mean? just checking we're on the same page

**Comment by aaime** [comment]
I believe we are not on the same page. The change in encoding happens because XStream + Jettison behavior changed, right?
I would expect them to be symmetric, if they encode something in a different way, they would also expect to read it back in that same different way. So if a REST automatic client is posting a JSON built the old way, I would expect to see an error, XStream + Jettison would not read it anymore. Does that make sense?

A deployed system is normally a complex beast, made of different components, built by different people (and sometimes different companies, on contract). Upgrading GeoServer (maybe to get a security fix?) should not require changing the other components of the deployment. If it does, then it should be clearly advertised in the docs (and discussed by the PSC).

**Comment by groldan** [comment]
> The change in encoding happens because XStream + Jettison behavior changed, right?

I'm not sure if you realize what I'm trying to do is to preserve the old behavior?

working on that `XStreamCatalogListConverter` issue which seems to be the only problematic one now

**Comment by aaime** [comment]
> > The change in encoding happens because XStream + Jettison behavior changed, right?
> 
> I'm not sure if you realize what I'm trying to do is to preserve the old behavior?

I am. But tests failing means there is currently a breakage in backward compatibility.

> 
> working on that `XStreamCatalogListConverter` issue which seems to be the only problematic one now

I was not aware it was limited to that converter, no previous message mentioned it. It's good news though, that's a converter under our control, so hopefully, it can be fixed.

File: src/rest/src/main/java/org/geoserver/rest/converters/XStreamCatalogListConverter.java [hunk 1]
```diff
@@ -216,7 +217,13 @@ public void encodeCollectionLink(String link, HierarchicalStreamWriter writer) {
 
         @Override
         protected XStream createXStreamInstance() {
-            return new SecureXStream(new JettisonMappedXmlDriver());
+            // needed for Jettison 1.4.1
```
Comment (aaime) [inline src/rest/src/main/java/org/geoserver/rest/converters/XStreamCatalogListConverter.java:220]: Ah ha!

**Comment by aaime** [comment]
All green, well done! I can't do a review right now, will try to look at this in the next few days.

**Comment by groldan** [comment]
Actually I'm trying to revert that test case to its original form and get exactly the same output as before out of `XStreamCatalogListConverter`. Added test cases to `CoverageControllerTest` for zero and single element lists, because those never returned unwrapped single-element results.

Problem is I either get
```
{"coverages":  {  "coverage": { "name": "tazdem"} } }
```
or 
```
{"coverages": [ {"coverage": {"name": "tazdem"} } ] }
```

But not the expected

```
{"coverages": {"coverage": [  { "name": "tazdem" } ] } }
```

**Comment by groldan** [comment]
ok, managed to preserve `XStreamCatalogListConverter`'s behavior as of 75925d37f87102f19aedd43d8fbfef63b0e4cfbc (replaces previous commit that changed the failing test instead, and adds a specific test case for the single-element case)

File: src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java [hunk 2]
```diff
@@ -310,6 +311,7 @@ public void testDirectoryXML() throws Exception {
     }
 
     @Test
+    @Ignore
```
Comment (mprins) [inline src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java:314]: Can you add a comment as to why this test should be ignored? 
```suggestion
    @Ignore("why is this ignored....")
```

**Review by mprins — COMMENTED** [review]
I see `boolean useSerializeAsArray = false;` in several classes and I'm wondering if it would be a good idea to move that into a public final in eg. `XStreamPersister.java` (not sure how that would work out dependency-wise

File: src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java (hunk 2 above)
Comment (groldan) [reply src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java:314]: I think it wasn't intentional, met me check

Comment (groldan) [reply src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java:314]: No, I missed that one. Was added by @aaime at 4bd66b2e222e464e03458892ca7f9737def3be00
I'll see what's going on

Comment (aaime) [reply src/restconfig/src/test/java/org/geoserver/rest/resources/ResourceControllerTest.java:314]: Yep it was part of the work in progress, that branch was not meant to be committed as-is

File: src/restconfig/src/test/java/org/geoserver/rest/catalog/DataStoreControllerTest.java [hunk 3]
```diff
@@ -72,6 +72,7 @@ public void testGetAsJSON() throws Exception {
     @Test
     public void testGetAllAsJSON() throws Exception {
         JSON json = getAsJSON(ROOT_PATH + "/workspaces/sf/datastores.json");
+        print(json);
```
Comment (aaime) [inline src/restconfig/src/test/java/org/geoserver/rest/catalog/DataStoreControllerTest.java:75]: I'd get rid of all the extra "print" in this PR.

**Review by aaime — COMMENTED** [review]
Looks pretty good in general! Agree with the two points raised by @mprins (ignored test, and sharing duplicated code if possible).

**Comment by aaime** [comment]
Looks like there are legit test failures (at one point I believe the build was green though? maybe I'm confusing it with something else...)

**Comment by groldan** [comment]
> Looks like there are legit test failures (at one point I believe the build was green though? maybe I'm confusing it with something else...)

It was green, but haven't seen that one that was `@Ignore`d. I'll get to it asap

**Comment by aaime** [comment]
Needs a bump to 1.4.19: https://github.com/geoserver/geoserver/pull/5636

**Comment by groldan** [comment]
just a heads up that I'm resuming work on this

**Comment by aaime** [comment]
How are things going @groldan ?

**Comment by groldan** [comment]
> How are things going @groldan ?

dancing with priorities @aaime, but I think I can finally get to this before the end of the week

**Comment by groldan** [comment]
Closing as superseded by #5730 with a single commit and corrected branch name.

**Pull request closed without merging**